lib/vector-config-macros/src/component_name.rs

```diff
@@ -1,22 +1,177 @@
+    // some code is trying to reference that struct or enum or what have you. This leads to a large
+    // number of errors that are technically related to the problem at hand, but ultimately are all
+    // solved by fixing the error with the macro usage.
+    //
+    // Derive macros function differently, such that the original item is always left alone and only
+    // new tokens are generated, so we can emit errors without causing more errors, but we've lost
+    // some of the semantic information about the original usage of `configurable_component` by the
+    // time we're here, running _this_ macro.
+    //
+    // To deal with this, we specifically look for component type-specific attributes that define
+    // the name, and higher up in `configurable_component`, we use the appropriate attribute for the
+    // component type at hand. This ends up giving output that follows this pattern:
+    //
+    // `#[configurable_component(source)]` -> `#[source_component]`
+    // `#[configurable_component(source("foo"))]` -> `#[source_component("foo")]`
+    //
+    // This allows us to determine the component type originally passed to `configurable_component`
+    // so that, in the case of the nameless example above, we can generate an error message attached
+    // to the span for `source`, such as: "sources must specify a name (e.g. `source("name")`)"
+    //
+    // Secondly, and finally, we always expect a single one of these helper attributes defining the
+    // name: no more, no less. Even though `configurable_component` should correctly adhere to that,
+    // we still need to go through the motions of verifying it here... which may capture either
+    // someone manually using the derive incorrectly, or an actual bug in our usage via other macros.
+    let mut errors = Vec::new();
+    let mut component_names = input
+        .attrs
+        .iter()
+        .filter_map(|attr| match attr_to_component_name(attr) {
+            Ok(component_name) => component_name,
+            Err(e) => {
+                errors.push(e);
+                None
+            }
+        })
+        .collect::<Vec<_>>();
+
+    if !errors.is_empty() {
+        let mut main_error = errors.remove(0);
+        for error in errors.drain(..) {
+            main_error.combine(error);
+        }
+
+        return main_error.into_compile_error().into();
+    }
+
+    // Any component names we have now have been validated, so we just need to check and make sure
+    // we actually have one, and only one, and spit out the correct errors otherwise.
+    if component_names.is_empty() {
+        return Error::new(
+            ident.span(),
+            "component must have a name defined (e.g. `#[component_name(\"foobar\")]`)",
+        )
+        .into_compile_error()
+        .into();
+    }
+
+    if component_names.len() > 1 {
+        return Error::new(ident.span(), "component cannot have multiple names defined")
```

Comment:
Presently, this would only matter for the purpose of generating example configurations... which is to say that if we supported an old/new name for transform via the normal pathways (i.e. `#[serde(alias = "...")]`) then a lack of setting multiple component names in `NamedComponent` would mean you could only generate an example configuration for the new name, not the old name.

We would still load configurations themselves, however, based on whatever aliases were present. That is obviously a divergence in behavior, although there aren't a ton of good solutions that come to mind, although it strikes me that not allowing people to use deprecated names in `vector generate` seems fine... because that's not load-bearing infrastructure, and as long as we don't break Vector loading real configs.... who cares?

Reply:
Ah, yes yes yes. I think that's good, and would prefer the `gen` subcommand to push people to the "correct" name regardless. My brain is still coming back up to speed and I totally forgot about `alias` 😓